### ISSUE TITLE
BigInt::to_dec_string is very slow and significant speedups are possible

### ISSUE THREAD

**Issue by guidovranken** [randombit/botan#2766]
```cpp
#include <boost/multiprecision/cpp_int.hpp>
#include <botan/bigint.h>
#include <sstream>

static int HexCharToDec(const char c) {
    if ( c >= '0' && c <= '9' ) {
        return c - '0';
    } else if ( c >= 'a' && c <= 'f' ) {
        return c - 'a' + 10;
    } else if ( c >= 'A' && c <= 'F' ) {
        return c - 'A' + 10;
    } else {
        assert(0);
    }
}
std::string HexToDec(std::string s) {
    std::string ret;
    bool negative = false;

    if ( s.empty() ) {
        return ret;
    }

    if ( s.size() >= 2 && s[0] == '0' && s[1] == 'x' ) {
        s = s.substr(2);
    }

    if ( s.size() >= 1 && s[0] == '-' ) {
        s = s.substr(1);
        negative = true;
    }

    boost::multiprecision::cpp_int total;

    for (long i = s.size() - 1; i >= 0; i--) {
        total += boost::multiprecision::cpp_int(HexCharToDec(s[i])) << ((s.size()-i-1)*4);
    }

    std::stringstream ss;
    if ( negative ) ss << "-";
    ss << total;

    if ( ss.str().empty() ) {
        return "0";
    } else {
        return ss.str();
    }
}

#if 1
static std::string to_dec(const Botan::BigInt& b) {
    return b.to_dec_string();
}
#else
static std::string to_dec(const Botan::BigInt& b) {
    return HexToDec(b.to_hex_string());
}
#endif

int main(void)
{
    for (size_t i = 0; i < 4000; i++) {
        Botan::BigInt b(1);
        b <<= i;
        printf("%s\n", to_dec(b).c_str());
    }
    return 0;
}
```

Using the alternative `to_dec` function that uses Boost results in a huge speedup (x60 in my benchmark, clang -O3 on x64).

I know this is a known issue (https://github.com/randombit/botan/blob/d26412c05270155debc459728bec9b7d47c78f5e/src/lib/math/bigint/bigint.h#L497) but the fact that my unoptimized conversion function (which I hacked together for Cryptofuzz) is so much faster suggests that a very significant speedup of `to_dec_string` is possible if you would optimize it.

**Comment by randombit** [comment]
The main reason for the slowdown is that the conversion is constant time, and ct divide is quite slow.

A more fair comparison would be BoringSSL's `BN_bn2dec` which similarly attempts to be ct. It looks like their code divides by either 1000000000 or 10000000000000000000 (depending on the word size) while we always do 100 (ie produce 2 digits per divide operation), so that's certainly something to explore.

We could also add a `to_dec_string_vartime` for use by Cryptofuzz or similar code where ct guarantees don't matter. [Or, maybe simpler (?), just make `to_dec_string` be vartime iff we are in unsafe-for-production-fuzzer-mode]

**Comment by guidovranken** [comment]
Ok, I just wasn't sure if you were aware of the unrealized performance gains.

I personally don't need a fast `to_dec_string` for functional purposes in Cryptofuzz: it works fine by converting to hex and then using Boost to convert to dec, and that's fast enough.

However, a (faster) `to_dec_string` can be useful to test whether conversion to decimal string always works as expected.

Out of curiosity why is `to_dec_string` constant-time?

**Comment by randombit** [comment]
> Out of curiosity why is to_dec_string constant-time?

Basically on the theory that it is better to CT everything viable rather than run the risk that something ends up being exploitable, and then we get to deal with patching and CVEs and etc. Whereas the worst that happens if something *is* CT is that it is too slow and someone opens an issue to complain. :) And then I have an opportunity to look for interesting new optimizations to make it both fast and CT.

**Comment by randombit** [comment]
#2768 goes to converting 19 digits per division instead of only 2, and as a result is 8-10x faster. [Probably only 4.5x faster on 32-bit since we are only converting 9 digits in that case]. I can replicate the time you see with Boost and even with this change Boost is still ~5x faster on my machine so you'll probably want to stick with that if this is a bottleneck for you. 99% of the runtime now is in the CT division routine, so making it any faster requires using a better algorithm.

**Comment by guidovranken** [comment]
Thanks! I will integrate support for both hex and dec string soon.

I've noticed that since https://github.com/randombit/botan/pull/2768 ```Botan::lcm(0, 0)``` now crashes:

```
terminate called after throwing an instance of 'Botan::Invalid_Argument'
  what():  ct_divide: cannot divide by zero
Aborted (core dumped)
```

whereas previously it would return 0. Is this intentional?

**Comment by randombit** [comment]
Not directly intentional though I did intentionally add a divide by zero check for ct_divide which had previously been missing. It seems LCM of (0,0) is not well defined? But I suppose if `a==b` then `LCM(a,b)` is arguably always `a` regardless of the value.

Turns out though that LCM function is literally only used in the context of RSA keygen and maybe it makes sense to just remove `lcm` as a public function and define it directly within `rsa.cpp`...

I just pushed some changes here to fix it in the short term but may remove this fn entirely for 3.0
